asio/CMakeLists.txt

```diff
@@ -0,0 +1,240 @@
+
+    if(Boost_FOUND)
+      include_directories(${Boost_INCLUDE_DIRS})
+      link_directories(${Boost_LIBRARY_DIRS})
+      list(APPEND CPPdefinitions BOOST_ALL_NO_LIB=1
+        BOOST_CHRONO_DONT_PROVIDE_HYBRID_ERROR_HANDLING
+      )
+      list(APPEND EXTRA_LIBS ${Boost_LIBRARIES})
+    endif()
+    #####################################################################
+  endif()
+
+
+  #######################################################################
+  # use openssl if found
+  set(OPENSSL_ROOT_DIR "/usr" CACHE PATH "path to the openssl lib")
+  find_package(OpenSSL HINTS ${OPENSSL_ROOT_DIR} /opt/local)
+  #######################################################################
+
+  if(OpenSSL_FOUND)
+    list(APPEND EXTRA_LIBS ${openSSL_LIBRARIES})
+    message(STATUS "HAVE_OPENSSL")
+    list(APPEND CPPdefinitions HAVE_OPENSSL)
+    include_directories(BEFORE ${openSSL_INCLUDE_DIR})
+  # else()
+  #   find_library(OpenSSL_LIBRARY ssl HINTS ${OpenSSL_ROOT_DIR}/lib)
+  #   find_library(CRYPTO_LIBRARY crypto HINTS ${OpenSSL_ROOT_DIR}/lib)
+  #   if(OpenSSL_LIBRARY AND CRYPTO_LIBRARY)
+  #     set(OpenSSL_FOUND ON)
+  #     message(STATUS "USE CRYPTO_LIBRARY")
+  #     list(APPEND CPPdefinitions HAVE_OPENSSL)
+  #     list(APPEND EXTRA_LIBS ${OpenSSL_LIBRARY} ${CRYPTO_LIBRARY})
+  #     include_directories(BEFORE ${OpenSSL_ROOT_DIR}/include)
+  #   endif()
+  endif()
+  #######################################################################
+else()
+  message(STATUS "ASIO USE CONAN_TARGETS")
+  if(CONAN_LIBS_OPENSSL)
+    list(APPEND CPPdefinitions HAVE_OPENSSL)
+    message(STATUS "have CONAN_LIBS_OPENSSL")
+  endif()
+endif()
+
+
+if(UNIX)
+  find_library(PTHREAD_LIBRARY pthread REQUIRED)
+  message(STATUS "USE ${PTHREAD_LIBRARY}")
+  list(APPEND EXTRA_LIBS ${PTHREAD_LIBRARY})
+endif()
+
+message(STATUS "CPPdefinitions:${CPPdefinitions}")
+
+
+# Installation configuration
+include(GNUInstallDirs)
+set(ConfigPackageLocation ${CMAKE_INSTALL_LIBDIR}/cmake/asio)
+
+# Interface library
+add_subdirectory(include)
+
+
+if($ENV{CONAN_RUN_TESTS})
+  set(ASIO_BUILD_TESTS ON CACHE BOOL "" FORCE)
+  message(STATUS "ASIO CONAN_RUN_TESTS")
+endif()
+
+
+if(ASIO_BUILD_EXAMPLES OR ASIO_BUILD_TESTS)
+  # enable ctest
+  enable_testing()
+
+  # Sources: examples, tests
+  add_subdirectory(src)
+endif()
+
+
+# Installation
+include(CMakePackageConfigHelpers)
+configure_package_config_file(asio-config.cmake.in
+  ${CMAKE_CURRENT_BINARY_DIR}/asio-config.cmake
+  INSTALL_DESTINATION ${ConfigPackageLocation})
+write_basic_package_version_file(
+  ${CMAKE_CURRENT_BINARY_DIR}/asio-config-version.cmake
+  VERSION ${asio_VERSION}
+  COMPATIBILITY SameMajorVersion)
+install(FILES ${CMAKE_CURRENT_BINARY_DIR}/asio-config-version.cmake
+  ${CMAKE_CURRENT_BINARY_DIR}/asio-config.cmake
+  DESTINATION ${ConfigPackageLocation})
+install(EXPORT asio-targets
+  DESTINATION ${ConfigPackageLocation}
+  NAMESPACE asio::)
```

Comment:
There are many things wrong with these install rules:

* Drop `configure_package_config_file`, you just need a simple config file that you install directly.
* You are missing install components. If a user vendors the library then it becomes impossible to tell this library's files apart from his. Since this is a header only library, everything can go in the `asio_Development` install component.
* Where is the include directory installed?

And some nitpicks:

* `write_basic_package_version_file` will pick up the version from the `project` command and the default output location of the file is already in the current build directory, they just add noise here.
* `ConfigPackageLocation` should be `asio_INSTALL_CMAKEDIR`, which is a string cache variable marked as advanced. Makes it trivial to relocate the config files for package managers like vcpkg.